### ISSUE TITLE
Too much CFLAGS parsed

### ISSUE THREAD

**Issue by fx-carton** [dense-analysis/ale#2510]
I updated ALE, and I got problems with CFLAGS again. I'm guessing it is caused by #2131, because some filenames (not options) are kept.
In my case, the command line included `-pedantic filename.c`, which are two arguments but are treated as one by the parser.
This is really a problem in my case, because for some reason having the input twice in the command line (`filename.c` and `< /tmp/.../filename.c`) makes gcc and thus vim hang. This makes ALE unusable, as vim hangs as soon as the file is open or modified.

It seems to me that a whitelist approach would be better than a blacklist.
I would suggest reverting that commit, and add `-isystem` as a known flag (which was the issue of that PR).

Here is the relevant part of `ALEInfo`:
```
(finished - exit code 0) ['/bin/bash', '-c', 'cd /home/user/somedir && make -n']
<<<OUTPUT STARTS>>>
cc -O2 -Wall -std=c99 -pedantic    prog.c libexample.a   -o prog
<<<OUTPUT ENDS>>>
(started) ['/bin/bash', '-c', '''gcc'' -S -x c -fsyntax-only -iquote ''/home/user/somedir'' -O2 -Wall -std=c99 -pedantic prog.c libexample.a -Wall -std=c89 -pedantic - < ''/tmp/vwPopcc/4/prog.c''']
```

As a note, it also seems that #2131 removed a test I added in #2168 that I think would have catched the issue.

I don't have a lot of free time right now, so I'll just downgrade ALE for the time being. If I get a chance to fix this, I'll make a PR.

**Comment by w0rp** [comment]
Parsing C compiler flags is incredibly complicated. The only practical solution is to continue to add more test cases until all of the important cases are covered. Please feel free to submit a pull request to fix this, which also supports other existing valid use cases.

**Comment by fx-carton** [comment]
I think a whitelist would allow parsing the different kinds of arguments correctly, by having several cases:
- the include flags (`-I`, `-isystem`) which have one argument that is a filename (and might need adjusting)
- some flags that have an argument which is not a filename (eg. `-D`)
- some flags without arguments (eg. warning flags, `-pedantic` and probably a lot more)

If I get some time to spend on this, I'll look at the arguments using `gcc`'s man page, filter the unwanted ones and classify the others into the categories listed above.

**Comment by w0rp** [comment]
Sure, let me know how it goes.

**Comment by fx-carton** [comment]
I do agree that it is extremely complicated and need more tests.

**Comment by w0rp** [comment]
Now fixed by your PR #2590. Thanks for that!
